README
NAME
    IO::CaptureOutput - capture STDOUT and STDERR from Perl code,
    subprocesses or XS

VERSION
    This documentation describes version 1.1102.

SYNOPSIS
         use IO::CaptureOutput qw(capture qxx qxy);
 
         # STDOUT and STDERR separately
         capture { noisy_sub(@args) } \$stdout, \$stderr;
 
         # STDOUT and STDERR together
         capture { noisy_sub(@args) } \$combined, \$combined;
 
         # STDOUT and STDERR from external command
         ($stdout, $stderr, $success) = qxx( @cmd );
 
         # STDOUT and STDERR together from external command
         ($combined, $success) = qxy( @cmd );

DESCRIPTION
    This module provides routines for capturing STDOUT and STDERR from perl
    subroutines, forked system calls (e.g. "system()", "fork()") and from XS
    or C modules.

FUNCTIONS
    The following functions will be exported on demand.

  capture()
         capture \&subroutine, \$stdout, \$stderr;

    Captures everything printed to "STDOUT" and "STDERR" for the duration of
    &subroutine. $stdout and $stderr are optional scalars that will contain
    "STDOUT" and "STDERR" respectively.

    "capture()" uses a code prototype so the first argument can be specified
    directly within brackets if desired.

         # shorthand with prototype
         capture { print __PACKAGE__ } \$stdout, \$stderr;

    Returns the return value(s) of &subroutine. The sub is called in the
    same context as "capture()" was called e.g.:

         @rv = capture { wantarray } ; # returns true
         $rv = capture { wantarray } ; # returns defined, but not true
         capture { wantarray };       # void, returns undef

    "capture()" is able to capture output from subprocesses and C code,
    which traditional "tie()" methods of output capture are unable to do.

    Note: "capture()" will only capture output that has been written or
    flushed to the filehandle.

    If the two scalar references refer to the same scalar, then "STDERR"
    will be merged to "STDOUT" before capturing and the scalar will hold the
    combined output of both.

         capture \&subroutine, \$combined, \$combined;

    Normally, "capture()" uses anonymous, temporary files for capturing
    output. If desired, specific file names may be provided instead as
    additional options.

         capture \&subroutine, \$stdout, \$stderr, $out_file, $err_file;

    Files provided will be clobbered, overwriting any previous data, but
    will persist after the call to "capture()" for inspection or other
    manipulation.

    By default, when no references are provided to hold STDOUT or STDERR,
    output is captured and silently discarded.

         # Capture STDOUT, discard STDERR
         capture \&subroutine, \$stdout;
 
         # Discard STDOUT, capture STDERR
         capture \&subroutine, undef, \$stderr;

    However, even when using "undef", output can be captured to specific
    files.

         # Capture STDOUT to a specific file, discard STDERR
         capture \&subroutine, \$stdout, undef, $outfile;
 
         # Discard STDOUT, capture STDERR to a specific file
         capture \&subroutine, undef, \$stderr, undef, $err_file;
 
         # Discard both, capture merged output to a specific file
         capture \&subroutine, undef, undef, $mergedfile;

    It is a fatal error to merge STDOUT and STDERR and request separate,
    specific files for capture.

         # ERROR:
         capture \&subroutine, \$stdout, \$stdout, $out_file, $err_file;
         capture \&subroutine, undef, undef, $out_file, $err_file;

    If either STDOUT or STDERR should be passed through to the terminal
    instead of captured, provide a reference to undef -- "\undef" -- instead
    of a capture variable.

         # Capture STDOUT, display STDERR
         capture \&subroutine, \$stdout, \undef;
 
         # Display STDOUT, capture STDERR
         capture \&subroutine, \undef, \$stderr;

  capture_exec()
         ($stdout, $stderr, $success, $exit_code) = capture_exec(@args);

    Captures and returns the output from "system(@args)". In scalar context,
    "capture_exec()" will return what was printed to "STDOUT". In list
    context, it returns what was printed to "STDOUT" and "STDERR" as well as
    a success flag and the exit value.

         $stdout = capture_exec('perl', '-e', 'print "hello world"');
 
         ($stdout, $stderr, $success, $exit_code) = 
             capture_exec('perl', '-e', 'warn "Test"');

    "capture_exec" passes its arguments to "system()" and on MSWin32 will
    protect arguments with shell quotes if necessary. This makes it a handy
    and slightly more portable alternative to backticks, piped "open()" and
    "IPC::Open3".

    The $success flag returned will be true if the command ran successfully
    and false if it did not (if the command could not be run or if it ran
    and returned a non-zero exit value). On failure, the raw exit value of
    the "system()" call is available both in the $exit_code returned and in
    the $? variable.

       ($stdout, $stderr, $success, $exit_code) = 
           capture_exec('perl', '-e', 'warn "Test" and exit 1');
 
       if ( ! $success ) {
           print "The exit code was " . ($exit_code >> 8) . "\n";
       }

    See perlvar for more information on interpreting a child process exit
    code.

  capture_exec_combined()
         ($combined, $success, $exit_code) = capture_exec_combined(
             'perl', '-e', 'print "hello\n"', 'warn "Test\n"
         );

    This is just like "capture_exec()", except that it merges "STDERR" with
    "STDOUT" before capturing output.

    Note: there is no guarantee that text printed to "STDOUT" and "STDERR"
    in the subprocess will be appear in order. The actual order will depend
    on how IO buffering is handled in the subprocess.

  qxx()
    This is an alias for "capture_exec()".

  qxy()
    This is an alias for "capture_exec_combined()".

SEE ALSO
    *   IPC::Open3

    *   IO::Capture

    *   IO::Utils

    *   IPC::System::Simple

AUTHORS
    *   Simon Flack <simonflk _AT_ cpan.org> (original author)

    *   David Golden <dagolden _AT_ cpan.org> (co-maintainer since version
        1.04)

COPYRIGHT AND LICENSE
    Portions copyright 2004, 2005 Simon Flack. Portions copyright 2007, 2008
    David Golden. All rights reserved.

    You may distribute under the terms of either the GNU General Public
    License or the Artistic License, as specified in the Perl README file.



Changes
Revision history for IO::CaptureOutput

{{$NEXT}}

1.1105    2019-10-25 17:40:04-04:00 America/New_York

    [META]

    - Add deprecation data to META file.

1.1104    2015-01-23 22:03:57-05:00 America/New_York

    [PREREQS]

    - Drop File::Spec 3.27 prereq except on MSWin32

1.1103    2014-02-27 19:54:14-05:00 America/New_York

    [FIXED]

    - Fixed rt.cpan.org #71775: test failure on non-threaded Win32 perls

    [CHANGED]

    - Updated metadata, repo and bugtracker

    - Switched distribution management to Dist::Zilla

1.1102 Mon Feb 15 07:36:53 EST 2010

    - Fixed: RT#45160 test that confused archname and osname
      (reported by Dolmen and Tux)

1.1101 Tue Mar  3 23:30:45 EST 2009

    - Fixed Makefile.PL; no functional changes

1.11 Mon Feb  9 21:47:37 EST 2009

    - Fixed: bug in creation of STDERR capture proxy object

1.1001 Tue Jan 20 18:32:27 EST 2009

    - Test fix: marked some wperl.exe tests as TODO on Windows Vista/Win7

1.10 Sun Sep 21 09:34:31 EDT 2008

    - Added: functions to capture output from external command now also
      return a success flag and the exit code

    - Documented: shorter, punchier synopsis

1.09 Thu Jul  3 09:17:35 EDT 2008

    - Added: parameter validation and error messages

    - Documented: examples of capturing only to a file and parameter errors

1.08_50 Mon May 19 23:19:01 EDT 2008

    - Changed: doesn't bother to read captured output back from file if no
      scalar reference is provided to receive it

    - Fixed: 'capture { .. } undef, undef, $capture_file' now correctly
      merges STDOUT and STDERR

1.0801 Sat Feb  9 09:41:11 EST 2008

    - testfix: skip some error handling tests if temp files can't be made
      read-only (e.g. running test as root)

    - added some prerequisites and meta information to Build.PL

    - added an 'examples' directory

1.08 Wed Feb  6 22:43:37 EST 2008

    - added additional tests and revised documentation

1.07_01 Wed Jan 30 20:57:15 EST 2008

    - added support for capturing to named files instead of anonymous
      temporary ones (adapted from patch from Anton Berezin)

    - added support to pass through STDOUT or STDERR without capture while
      capturing the other one (adapted from patch from Anton Berezin)

1.0601 Sun Dec 30 17:31:31 EST 2007

    - relaxed capture_exec.t to avoid failing under noisy perl configurations

1.06 Mon, 26 Nov 2007 16:40:40 EST

    - no changes from 1.05_53

1.05_53 Sat Nov 17 18:45:08 EST 2007

    - MSVC++ fixes for Inline::C testing breaks on other compilers; this
      version uses separate C routines for MSVC++ and other compilers

1.05_52 Sat Nov 17 14:47:52 EST 2007

    - changed Inline::C code for printing to stdout to match that for stderr;
      works around a bug if the intalled MSVC++ version doesn't match the
      version used to build ActiveState Perl on MSWin32.

1.05_51 Sat Nov 17 13:36:39 EST 2007

    - changed Inline::C code for printing to stderr to attempt to work around
      a segfault bug on some MSWin32 systems (Alternative described by
      BrowserUK at http://www.perlmonks.org/?node_id=649021 )

    - minor Perl::Critic cleanup to pass xt tests

1.05 Fri Nov 16 08:44:06 EST 2007

    *** No changes since 1.04_03; key changes since 1.03 follows ***

    - added support for capture() merging STDERR and STDOUT to the same
      scalar, i.e. "capture( \&foo, \$out, \$out )" (RT#24695)

    - added capture_exec_combined() and alias qxy() to merge STDERR and STDOUT
      for subprocess execution (RT#29289)

    - bugfix: stop crashing if output filehandles are previously closed or
      if running without a console on MSWin32 (i.e. wperl.exe) (RT#23676)

    - bugfix: capture.t would fail on Win32 if Inline::C was installed;
      fixed the sample C code used in testing to flush stderr (RT#21712)

    - bugfix: was failing on perl 5.8.0 due to extraneous space in call
      to open(); removing the space fixed the bug (RT#22044)

1.04_03 Sun Nov  4 11:35:57 EST 2007

    - suppress warnings included in captured output when console output
      handles were previously closed; only affects some versions/builds of
      perl on some platforms (found by Slaven Rezic 5.6.2 on FreeBSD 6.2)

1.04_02 Wed Oct 31 00:52:49 EDT 2007

    - added support for capture() merging STDERR and STDOUT to the same
      scalar, i.e. "capture( \&foo, \$out, \$out )" (RT#24695)

    - added capture_exec_combined() and alias qxy() to merge STDERR and STDOUT
      for subprocess execution (RT#29289)

    - bugfix: stop crashing if output filehandles are previously closed or
      if running without a console on MSWin32 (i.e. wperl.exe) (RT#23676)

    - cleaned up documentation for consistency

1.04_01 Mon Oct 29 22:56:45 EDT 2007

    - DAGOLDEN added as maintainer; updated build system and packaging for
      ongoing maintenance

    - bugfix: capture.t would fail on Win32 if Inline::C was installed;
      fixed the sample C code used in testing to flush stderr (RT#21712)

    - bugfix: was failing on perl 5.8.0 due to extraneous space in call
      to open(); removing the space fixed the bug (RT#22044)

1.03  2005-03-25
    - Add t/pod_coverage.t

1.02  2004-11-22
    - add extra quoting on Win32 platform
    - Win32 won't unlink a file that is open in write mode, so close it first
    - minor POD tweaks

1.0  2004-11-21
    - Initial revision


Build.PL
use strict;
use lib 'inc';
use Module::Build::DAGOLDEN;

my $build = Module::Build::DAGOLDEN->new( 
    module_name         => 'IO::CaptureOutput',
    dist_author         => 'David A. Golden <dagolden@cpan.org>',
    license             => 'perl',
    create_readme       => 1,
    create_makefile_pl  => 'traditional',
    requires => {
        'Carp' => 0, 
        'Exporter' => 0, 
        'File::Basename' => 0, 
        'File::Temp' => 0.16,  # many bug fixes
        'Symbol' => 0, 
    },
    build_requires => {
        'File::Spec' => 3.27, # many bug fixes
        'Test::More' => 0.62,  # newer API
        'IO::File' => 0, 
    },
    add_to_cleanup => [ '_Inline' ],
    meta_merge            => { 
        no_index => { 
            directory => [ qw{
                examples
                inc
            }],
        },
        resources => {
            repository => 'git://github.com/dagolden/io-captureoutput.git'
        },
    },
);

$build->create_build_script;



Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "(DEPRECATED) capture STDOUT and STDERR from Perl code, subprocesses or XS",
  "AUTHOR" => "Simon Flack <simonflk\@cpan.org>, David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "IO-CaptureOutput",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "IO::CaptureOutput",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "File::Basename" => 0,
    "File::Temp" => "0.16",
    "Symbol" => 0,
    "strict" => 0,
    "vars" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => "3.27",
    "IO::File" => 0,
    "Test::More" => "0.62"
  },
  "VERSION" => "1.1106",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Spec" => "3.27",
  "File::Temp" => "0.16",
  "IO::File" => 0,
  "Symbol" => 0,
  "Test::More" => "0.62",
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

if ( $^O eq 'MSWin32' ) {
	$WriteMakefileArgs{PREREQ_PM}{'File::Spec'} = $FallbackPrereqs{'File::Spec'} = '3.27';
}

WriteMakefile(%WriteMakefileArgs);


lib/IO/CaptureOutput.pm
use strict;
use warnings;

package IO::CaptureOutput;
# ABSTRACT: (DEPRECATED) capture STDOUT and STDERR from Perl code, subprocesses or XS

our $VERSION = '1.1106';

use vars qw/@ISA @EXPORT_OK %EXPORT_TAGS $CarpLevel/;
use Exporter;
use Carp qw/croak/;
@ISA = 'Exporter';
@EXPORT_OK = qw/capture capture_exec qxx capture_exec_combined qxy/;
%EXPORT_TAGS = (all => \@EXPORT_OK);
$CarpLevel = 0; # help capture report errors at the right level

sub _capture (&@) { ## no critic
    my ($code, $output, $error, $output_file, $error_file) = @_;

    # check for valid combinations of input
    {
      local $Carp::CarpLevel = 1;
      my $error = _validate($output, $error, $output_file, $error_file);
      croak $error if $error;
    }

    # if either $output or $error are defined, then we need a variable for
    # results; otherwise we only capture to files and don't waste memory
    if ( defined $output || defined $error ) {
      for ($output, $error) {
          $_ = \do { my $s; $s = ''} unless ref $_;
          $$_ = '' if $_ != \undef && !defined($$_);
      }
    }

    # merge if same refs for $output and $error or if both are undef --
    # i.e. capture \&foo, undef, undef, $merged_file
    # this means capturing into separate files *requires* at least one
    # capture variable
    my $should_merge =
      (defined $error && defined $output && $output == $error) ||
      ( !defined $output && !defined $error ) ||
      0;

    my ($capture_out, $capture_err);

    # undef means capture anonymously; anything other than \undef means
    # capture to that ref; \undef means skip capture
    if ( !defined $output || $output != \undef ) {
        $capture_out = IO::CaptureOutput::_proxy->new(
            'STDOUT', $output, undef, $output_file
        );
    }
    if ( !defined $error || $error != \undef ) {
        $capture_err = IO::CaptureOutput::_proxy->new(
            'STDERR', $error, ($should_merge ? 'STDOUT' : undef), $error_file
        );
    }

    # now that output capture is setup, call the subroutine
    # results get read when IO::CaptureOutput::_proxy objects go out of scope
    &$code();
}

# Extra indirection for symmetry with capture_exec, etc.  Gets error reporting
# to the right level
sub capture (&@) { ## no critic
    return &_capture;
}

sub capture_exec {
    my @args = @_;
    my ($output, $error);
    my $exit = _capture sub { system _shell_quote(@args) }, \$output, \$error;
    my $success = ($exit == 0 ) ? 1 : 0 ;
    $? = $exit;
    return wantarray ? ($output, $error, $success, $exit) : $output;
}

*qxx = \&capture_exec;

sub capture_exec_combined {
    my @args = @_;
    my $output;
    my $exit = _capture sub { system _shell_quote(@args) }, \$output, \$output;
    my $success = ($exit == 0 ) ? 1 : 0 ;
    $? = $exit;
    return wantarray ? ($output, $success, $exit) : $output;
}

*qxy = \&capture_exec_combined;

# extra quoting required on Win32 systems
*_shell_quote = ($^O =~ /MSWin32/) ? \&_shell_quote_win32 : sub {@_};
sub _shell_quote_win32 {
    my @args;
    for (@_) {
        if (/[ \"]/) { # TODO: check if ^ requires escaping
            (my $escaped = $_) =~ s/([\"])/\\$1/g;
            push @args, '"' . $escaped . '"';
            next;
        }
        push @args, $_
    }
    return @args;
}

# detect errors and return an error message or empty string;
sub _validate {
    my ($output, $error, $output_file, $error_file) = @_;

    # default to "ok"
    my $msg = q{};

    # \$out, \$out, $outfile, $errfile
    if (    defined $output && defined $error
        &&  defined $output_file && defined $error_file
        &&  $output == $error
        &&  $output != \undef
        &&  $output_file ne $error_file
    ) {
      $msg = "Merged STDOUT and STDERR, but specified different output and error files";
    }
    # undef, undef, $outfile, $errfile
    elsif ( !defined $output && !defined $error
        &&  defined $output_file && defined $error_file
        &&  $output_file ne $error_file
    ) {
      $msg = "Merged STDOUT and STDERR, but specified different output and error files";
    }

    return $msg;
}

# Captures everything printed to a filehandle for the lifetime of the object
# and then transfers it to a scalar reference
package IO::CaptureOutput::_proxy;
use File::Temp 0.16 'tempfile';
use File::Basename qw/basename/;
use Symbol qw/gensym qualify qualify_to_ref/;
use Carp;

sub _is_wperl { $^O eq 'MSWin32' && basename($^X) eq 'wperl.exe' }

sub new {
    my $class = shift;
    my ($orig_fh, $capture_var, $merge_fh, $capture_file) = @_;
    $orig_fh       = qualify($orig_fh);         # e.g. main::STDOUT
    my $fhref = qualify_to_ref($orig_fh);  # e.g. \*STDOUT

    # Duplicate the filehandle
    my $saved_fh;
    {
        no strict 'refs'; ## no critic - needed for 5.005
        if ( defined fileno($orig_fh) && ! _is_wperl() ) {
            $saved_fh = gensym;
            open $saved_fh, ">&$orig_fh" or croak "Can't redirect <$orig_fh> - $!";
        }
    }

    # Create replacement filehandle if not merging
    my ($newio_fh, $newio_file);
    if ( ! $merge_fh ) {
        $newio_fh = gensym;
        if ($capture_file) {
            $newio_file = $capture_file;
        } else {
            (undef, $newio_file) = tempfile;
        }
        open $newio_fh, "+>$newio_file" or croak "Can't write temp file for $orig_fh - $!";
    }
    else {
        $newio_fh = qualify($merge_fh);
    }

    # Redirect (or merge)
    {
        no strict 'refs'; ## no critic -- needed for 5.005
        open $fhref, ">&".fileno($newio_fh) or croak "Can't redirect $orig_fh - $!";
    }

    bless [$$, $orig_fh, $saved_fh, $capture_var, $newio_fh, $newio_file, $capture_file], $class;
}

sub DESTROY {
    my $self = shift;

    my ($pid, $orig_fh, $saved_fh, $capture_var, $newio_fh,
      $newio_file, $capture_file) = @$self;
    return unless $pid eq $$; # only cleanup in the process that is capturing

    # restore the original filehandle
    my $fh_ref = Symbol::qualify_to_ref($orig_fh);
    select((select ($fh_ref), $|=1)[0]);
    if (defined $saved_fh) {
        open $fh_ref, ">&". fileno($saved_fh) or croak "Can't restore $orig_fh - $!";
    }
    else {
        close $fh_ref;
    }

    # transfer captured data to the scalar reference if we didn't merge
    # $newio_file is undef if this file handle is merged to another
    if (ref $capture_var && $newio_file) {
        # some versions of perl complain about reading from fd 1 or 2
        # which could happen if STDOUT and STDERR were closed when $newio
        # was opened, so we just squelch warnings here and continue
        local $^W;
        seek $newio_fh, 0, 0;
        $$capture_var = do {local $/; <$newio_fh>};
    }
    close $newio_fh if $newio_file;

    # Cleanup
    return unless defined $newio_file && -e $newio_file;
    return if $capture_file; # the "temp" file was explicitly named
    unlink $newio_file or carp "Couldn't remove temp file '$newio_file' - $!";
}

1;

__END__

=pod

=head1 NAME

=head1 SYNOPSIS

    use IO::CaptureOutput qw(capture qxx qxy);

    # STDOUT and STDERR separately
    capture { noisy_sub(@args) } \$stdout, \$stderr;

    # STDOUT and STDERR together
    capture { noisy_sub(@args) } \$combined, \$combined;

    # STDOUT and STDERR from external command
    ($stdout, $stderr, $success) = qxx( @cmd );

    # STDOUT and STDERR together from external command
    ($combined, $success) = qxy( @cmd );

=head1 DESCRIPTION

B<This module is no longer recommended by the maintainer> - see
L<Capture::Tiny> instead.

This module provides routines for capturing STDOUT and STDERR from perl
subroutines, forked system calls (e.g. C<system()>, C<fork()>) and from XS
or C modules.

=head1 FUNCTIONS

The following functions will be exported on demand.

=head2 capture()

    capture \&subroutine, \$stdout, \$stderr;

Captures everything printed to C<STDOUT> and C<STDERR> for the duration of
C<&subroutine>. C<$stdout> and C<$stderr> are optional scalars that will
contain C<STDOUT> and C<STDERR> respectively.

C<capture()> uses a code prototype so the first argument can be specified
directly within brackets if desired.

    # shorthand with prototype
    capture C< print __PACKAGE__ > \$stdout, \$stderr;

Returns the return value(s) of C<&subroutine>. The sub is called in the
same context as C<capture()> was called e.g.:

    @rv = capture C< wantarray > ; # returns true
    $rv = capture C< wantarray > ; # returns defined, but not true
    capture C< wantarray >;       # void, returns undef

C<capture()> is able to capture output from subprocesses and C code, which
traditional C<tie()> methods of output capture are unable to do.

B<Note:> C<capture()> will only capture output that has been written or
flushed to the filehandle.

If the two scalar references refer to the same scalar, then C<STDERR> will
be merged to C<STDOUT> before capturing and the scalar will hold the
combined output of both.

    capture \&subroutine, \$combined, \$combined;

Normally, C<capture()> uses anonymous, temporary files for capturing
output.  If desired, specific file names may be provided instead as
additional options.

    capture \&subroutine, \$stdout, \$stderr, $out_file, $err_file;

Files provided will be clobbered, overwriting any previous data, but will
persist after the call to C<capture()> for inspection or other
manipulation.

By default, when no references are provided to hold STDOUT or STDERR,
output is captured and silently discarded.

    # Capture STDOUT, discard STDERR
    capture \&subroutine, \$stdout;

    # Discard STDOUT, capture STDERR
    capture \&subroutine, undef, \$stderr;

However, even when using C<undef>, output can be captured to specific
files.

    # Capture STDOUT to a specific file, discard STDERR
    capture \&subroutine, \$stdout, undef, $outfile;

    # Discard STDOUT, capture STDERR to a specific file
    capture \&subroutine, undef, \$stderr, undef, $err_file;

    # Discard both, capture merged output to a specific file
    capture \&subroutine, undef, undef, $mergedfile;

It is a fatal error to merge STDOUT and STDERR and request separate,
specific files for capture.

    # ERROR:
    capture \&subroutine, \$stdout, \$stdout, $out_file, $err_file;
    capture \&subroutine, undef, undef, $out_file, $err_file;

If either STDOUT or STDERR should be passed through to the terminal instead
of captured, provide a reference to undef -- C<\undef> -- instead of a
capture variable.

    # Capture STDOUT, display STDERR
    capture \&subroutine, \$stdout, \undef;

    # Display STDOUT, capture STDERR
    capture \&subroutine, \undef, \$stderr;

=head2 capture_exec()

    ($stdout, $stderr, $success, $exit_code) = capture_exec(@args);

Captures and returns the output from C<system(@args)>. In scalar context,
C<capture_exec()> will return what was printed to C<STDOUT>. In list
context, it returns what was printed to C<STDOUT> and C<STDERR> as well as
a success flag and the exit value.

    $stdout = capture_exec('perl', '-e', 'print "hello world"');

    ($stdout, $stderr, $success, $exit_code) =
        capture_exec('perl', '-e', 'warn "Test"');

C<capture_exec> passes its arguments to C<system()> and on MSWin32 will
protect arguments with shell quotes if necessary.  This makes it a handy
and slightly more portable alternative to backticks, piped C<open()> and
C<IPC::Open3>.

The C<$success> flag returned will be true if the command ran successfully
and false if it did not (if the command could not be run or if it ran and
returned a non-zero exit value).  On failure, the raw exit value of the
C<system()> call is available both in the C<$exit_code> returned and in the
C<$?> variable.

  ($stdout, $stderr, $success, $exit_code) =
      capture_exec('perl', '-e', 'warn "Test" and exit 1');

  if ( ! $success ) {
      print "The exit code was " . ($exit_code >> 8) . "\n";
  }

See L<perlvar> for more information on interpreting a child process exit
code.

=head2 capture_exec_combined()

    ($combined, $success, $exit_code) = capture_exec_combined(
        'perl', '-e', 'print "hello\n"', 'warn "Test\n"
    );

This is just like C<capture_exec()>, except that it merges C<STDERR> with
C<STDOUT> before capturing output.

B<Note:> there is no guarantee that text printed to C<STDOUT> and C<STDERR>
in the subprocess will be appear in order. The actual order will depend on
how IO buffering is handled in the subprocess.

=head2 qxx()

This is an alias for C<capture_exec()>.

=head2 qxy()

This is an alias for C<capture_exec_combined()>.

=head1 SEE ALSO

=for :list
* L<Capture::Tiny>
* L<IPC::Open3>
* L<IO::Capture>
* L<IO::Utils>
* L<IPC::System::Simple>

=cut
